### PR TITLE
docs(example): example extended with fabric

### LINKED ISSUE #362: docs(examples): extend supply chain app with fabric specific elements
### Description

Depends on #361, #360, #319 and #320 #347 

The supply chain app example should have a third ledger (Fabric) included that has a third smart contract on it that implements a shipment tracking feature for the existing example application where we can only register harvests but not actually track the shipments themselves.
The model should be updated so that bookshelves are referencing a shipment not a harvest and shipments reference a harvest so that the reference chain is complete from shelf to the lumberjack's chainsaw buzzing in the forest.

### Acceptance Criteria
1. Code compiles and tests are passing
2. The UI (`supply-chain-frontend` package) has a CRUD screen where one can add new shipments, associating them with harvests
2.1. Creating a shipment stores the data of said shipment on the contract's state in the Fabric ledger
3. The bookshelves and harvests CRUD screens is updated according to the model changes described above

cc: @sfuji822 @takeutak @jonathan-m-hamilton

### PR DESCRIPTION
**Pull request by jordigiam**
Resolve #362  
Depends on #616 
Depends on #634

### CODE REVIEW

**Review by takeutak — DISMISSED** [review]

File: examples/cactus-example-supply-chain-backend/src/main/typescript/supply-chain-app.ts [hunk 1]
```diff
@@ -200,6 +235,57 @@ export class SupplyChainApp {
     registryB.add(quorumConnector);
 
     await this.startNode(httpApiB, httpGuiB, registryB);
+
+    this.log.info(`Configuring Cactus Node for Ledger B...`);
```
Comment (petermetz) [inline examples/cactus-example-supply-chain-backend/src/main/typescript/supply-chain-app.ts:239]: @jordigiam Super minor nit pick, apologies: I'm guessing that was going to be `Ledger C` not `B`

File: examples/cactus-example-supply-chain-business-logic-plugin/src/main/typescript/business-logic-plugin/supply-chain-cactus-plugin.ts [hunk 2]
```diff
@@ -71,7 +91,8 @@ export class SupplyChainCactusPlugin
       contractAddress: this.options.contracts.bambooHarvestRepository.address,
       contractAbi: this.options.contracts.bambooHarvestRepository.abi,
       apiClient: this.options.quorumApiClient,
-      web3SigningCredential: this.options.web3SigningCredential,
+      web3SigningCredential: this.options
+        .web3SigningCredential as Web3SigningCredential,
```
Comment (petermetz) [inline examples/cactus-example-supply-chain-business-logic-plugin/src/main/typescript/business-logic-plugin/supply-chain-cactus-plugin.ts:95]: @jordigiam I'm guessing the casting here was necessary because the options object has the `web3SigningCredential` property declared as optional. Could it be that it's more fitting to change that from optional to mandatory and then no cast is necessary (plus the caller who passes in the options object is informed that way that we do assume the credentials to be present)

File: examples/cactus-example-supply-chain-business-logic-plugin/src/main/typescript/business-logic-plugin/supply-chain-cactus-plugin.ts [hunk 3]
```diff
@@ -88,7 +109,8 @@ export class SupplyChainCactusPlugin
       contractAddress: this.options.contracts.bookshelfRepository.address,
       contractAbi: this.options.contracts.bookshelfRepository.abi,
       besuApi: this.options.besuApiClient,
-      web3SigningCredential: this.options.web3SigningCredential,
+      web3SigningCredential: this.options
```
Comment (petermetz) [inline examples/cactus-example-supply-chain-business-logic-plugin/src/main/typescript/business-logic-plugin/supply-chain-cactus-plugin.ts:112]: @jordigiam I'm guessing the casting here was necessary because the options object has the `web3SigningCredential` property declared as optional. Could it be that it's more fitting to change that from optional to mandatory and then no cast is necessary (plus the caller who passes in the options object is informed that way that we do assume the credentials to be present)

File: examples/cactus-example-supply-chain-business-logic-plugin/src/main/typescript/business-logic-plugin/web-services/insert-shipment-endpoint-constants.ts [hunk 4]
```diff
@@ -0,0 +1,8 @@
+export class InsertShipmentEndpoint {
```
Comment (petermetz) [inline examples/cactus-example-supply-chain-business-logic-plugin/src/main/typescript/business-logic-plugin/web-services/insert-shipment-endpoint-constants.ts:1]: @jordigiam These constants classes will disappear completely eventually in favor of pulling the values directly from the openapi.json files. (just FYI, I'm okay to accept this as is for now)

File: examples/cactus-example-supply-chain-frontend/package.json [hunk 5]
```diff
@@ -47,6 +48,7 @@
     "@types/jasmine": "~3.5.0",
     "@types/jasminewd2": "~2.0.3",
     "@types/node": "^12.11.1",
+    "@types/ssh2": "0.5.46",
```
Comment (petermetz) [inline examples/cactus-example-supply-chain-frontend/package.json:51]: @jordigiam Did this turn out to be really necessary or was that compilation error only happening prior to running `npm run build:dev` (which produces the builds for the browser platform as well)

**Review by petermetz — CHANGES_REQUESTED** [review]

File: examples/cactus-example-supply-chain-backend/src/main/typescript/supply-chain-app.ts (hunk 1 above)
Comment (jordigiam) [reply examples/cactus-example-supply-chain-backend/src/main/typescript/supply-chain-app.ts:239]: Done

File: examples/cactus-example-supply-chain-business-logic-plugin/src/main/typescript/business-logic-plugin/supply-chain-cactus-plugin.ts (hunk 2 above)
Comment (jordigiam) [reply examples/cactus-example-supply-chain-business-logic-plugin/src/main/typescript/business-logic-plugin/supply-chain-cactus-plugin.ts:95]: If I change that from optional to mandatory the Fabric busines logic with the InsertShipmentEndpoint doesn't work, because the Fabric Ledger doesn't have the Web3SigningCredential. For now, I will keep the casting.

File: examples/cactus-example-supply-chain-frontend/package.json (hunk 5 above)
Comment (jordigiam) [reply examples/cactus-example-supply-chain-frontend/package.json]: Done

File: examples/cactus-example-supply-chain-business-logic-plugin/src/main/typescript/business-logic-plugin/web-services/insert-shipment-endpoint-constants.ts (hunk 4 above)
Comment (jordigiam) [reply examples/cactus-example-supply-chain-business-logic-plugin/src/main/typescript/business-logic-plugin/web-services/insert-shipment-endpoint-constants.ts]: Done

File: examples/cactus-example-supply-chain-business-logic-plugin/src/main/typescript/business-logic-plugin/supply-chain-cactus-plugin.ts [hunk 6]
```diff
@@ -38,7 +51,7 @@ export class SupplyChainCactusPlugin
   private readonly log: Logger;
   private readonly instanceId: string;
 
-  public get className() {
+  public get className(): string {
```
Comment (petermetz) [inline examples/cactus-example-supply-chain-business-logic-plugin/src/main/typescript/business-logic-plugin/supply-chain-cactus-plugin.ts:54]: @jordigiam Thanks for going the extra mile to fix linter warnings!

File: examples/cactus-example-supply-chain-business-logic-plugin/src/main/typescript/business-logic-plugin/supply-chain-cactus-plugin.ts (hunk 2 above)
Comment (petermetz) [reply examples/cactus-example-supply-chain-business-logic-plugin/src/main/typescript/business-logic-plugin/supply-chain-cactus-plugin.ts:84]: Oh, my bad, forgot that we use the same plugin for all of the ledgers, apologies!

**Review by petermetz — DISMISSED** [review]

**Comment by petermetz** [comment]
@jordigiam CI is failing due to it running out of disk space[1] which I remedied successfully in the past by adding this to the test where it gives out, same thing as here:
`packages/cactus-plugin-ledger-connector-fabric/src/test/typescript/integration/fabric-v2-2-x/run-transaction-endpoint-v1.test.ts`

```typescript
  // Always set to true when GitHub Actions is running the workflow.
  // You can use this variable to differentiate when tests are being run locally or by GitHub Actions.
  // @see https://docs.github.com/en/actions/reference/environment-variables
  if (process.env.GITHUB_ACTIONS === "true") {
    // Github Actions started to run out of disk space recently so we have this
    // hack here to attempt to free up disk space when running inside a VM of
    // the CI system.
    await Containers.pruneDockerResources();
  }
```


[1]:

```sh
ail out! packages/cactus-plugin-ledger-connector-fabric/src/test/typescript/integration/fabric-v1-4-x/run-transaction-endpoint-v1.test.ts
internal/fs/utils.js:220
    throw err;
    ^

Error: ENOSPC: no space left on device, write
    at Object.writeSync (fs.js:581:3)
    at Object.writeFileSync (fs.js:1275:26)
    at NYC.writeProcessIndex (/home/runner/work/cactus/cactus/node_modules/nyc/index.js:490:8)
    at /home/runner/work/cactus/cactus/node_modules/nyc/bin/nyc.js:68:9
    at ChildProcess.<anonymous> (/home/runner/work/cactus/cactus/node_modules/foreground-child/index.js:52:5)
    at ChildProcess.emit (events.js:210:5)
    at ChildProcess.EventEmitter.emit (domain.js:476:20)
    at maybeClose (internal/child_process.js:1021:16)
    at Process.ChildProcess._handle.onexit (internal/child_process.js:283:5) {
  errno: -28,
  syscall: 'write',
  code: 'ENOSPC'
}
```

**Review by takeutak — APPROVED** [review]

**Review by petermetz — CHANGES_REQUESTED** [review]
@jordigiam @AzaharaC Looks legit. 

TLDR: 
I made some fixes/chores/docs changes, is it okay to force push? Didn't want to accidentally erase your work.

Story:
When testing locally, the docker image build wasn't working for me because I broke it in a PR of mine a few weeks back (corda connector) so I added a hotfix for that in a commit that I recommend to keep separate from the main PR commit (so the PR would have 2 commits which is fine in this case IMO because I did not want to complicate things even further by opening a separate PR with the hot-fix and then have this one wait for that and be forced to depend on multiple parent branches, etc...) 

The other thing I did was publishing the container image to DockerHub under a new tag and updated the READMEs to use that image so that once we merge this PR, everyone will see the latest example in the root readme with Fabric included.

I have all the above on a branch on my work because I did not want to risk force pushing onto this and then erasing something that you were working on (but if you confirm that it's cool then I'll go ahead and perform the force push).

https://github.com/petermetz/cactus/tree/feature/extend-example-fabric

**Comment by AzaharaC** [comment]
> @jordigiam @AzaharaC Looks legit.
> 
> TLDR:
> I made some fixes/chores/docs changes, is it okay to force push? Didn't want to accidentally erase your work.
> 
> Story:
> When testing locally, the docker image build wasn't working for me because I broke it in a PR of mine a few weeks back (corda connector) so I added a hotfix for that in a commit that I recommend to keep separate from the main PR commit (so the PR would have 2 commits which is fine in this case IMO because I did not want to complicate things even further by opening a separate PR with the hot-fix and then have this one wait for that and be forced to depend on multiple parent branches, etc...)
> 
> The other thing I did was publishing the container image to DockerHub under a new tag and updated the READMEs to use that image so that once we merge this PR, everyone will see the latest example in the root readme with Fabric included.
> 
> I have all the above on a branch on my work because I did not want to risk force pushing onto this and then erasing something that you were working on (but if you confirm that it's cool then I'll go ahead and perform the force push).
> 
> https://github.com/petermetz/cactus/tree/feature/extend-example-fabric

@petermetz I have been reviewing your changes and I think that the push can be done without problems, so do it.

**Review by petermetz — APPROVED** [review]

**Comment by petermetz** [comment]
@takeutak Review was dismissed because of conflict resolution, could you please approve again? Thank you in advance!

**Review by takeutak — APPROVED** [review]
